core/unit_test/TestHalfOperators.hpp

```diff
@@ -0,0 +1,186 @@
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef TESTHALFOPERATOR_HPP_
+#define TESTHALFOPERATOR_HPP_
+namespace Test {
+#define FP16_EPSILON 0.0009765625F
+#define FP16_MAX 65504.0F           // Maximum normalized number
+#define FP16_MIN 0.000000059604645F // Minimum normalized positive half precision number
+using half_t = Kokkos::Experimental::half_t;
+using ViewType = Kokkos::View<double*>;
+enum {
+  UNA,
+  UNS,
+  ADD,
+  SUB,
+  MUL,
+  DIV,
+  NEG,
+  AND,
+  OR,
+  EQ,
+  NEQ,
+  LT,
+  GT,
+  LE,
+  GE,
+  N_OPS
+} op_types_e;
+
+struct Functor_TestHalfOperators {
+  half_t lhs, rhs;
+  double d_lhs, d_rhs;
+  ViewType actual_lhs, expected_lhs;
+
+  Functor_TestHalfOperators():
+    lhs(0),
+    rhs(0)
+  {
+    actual_lhs = ViewType("actual_lhs", N_OPS);
+    expected_lhs = ViewType("expected_lhs", N_OPS);
+    d_lhs = Kokkos::Experimental::cast_from_half<double>(lhs);
+    d_rhs = Kokkos::Experimental::cast_from_half<double>(rhs);
+  }
+
+  Functor_TestHalfOperators(half_t lhs, half_t rhs):
+    lhs(lhs),
+    rhs(rhs)
+  {
+    actual_lhs = ViewType("actual_lhs", N_OPS);
+    expected_lhs = ViewType("expected_lhs", N_OPS);
+    d_lhs = Kokkos::Experimental::cast_from_half<double>(lhs);
+    d_rhs = Kokkos::Experimental::cast_from_half<double>(rhs);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int &i) const {
+    actual_lhs(UNA) = Kokkos::Experimental::cast_from_half<double>(+lhs);
+    expected_lhs(UNA) = +d_lhs;
+
+    actual_lhs(UNS) = Kokkos::Experimental::cast_from_half<double>(-lhs);
+    expected_lhs(UNS) = -d_lhs;
+
+    actual_lhs(ADD) = Kokkos::Experimental::cast_from_half<double>(lhs + rhs);
+    expected_lhs(ADD) = d_lhs + d_rhs;
+
+    actual_lhs(ADD) = Kokkos::Experimental::cast_from_half<double>(lhs + rhs);
+    expected_lhs(ADD) = d_lhs + d_rhs;
+
+    actual_lhs(SUB) = Kokkos::Experimental::cast_from_half<double>(lhs - rhs);
+    expected_lhs(SUB) = d_lhs - d_rhs;
+
+    actual_lhs(MUL) = Kokkos::Experimental::cast_from_half<double>(lhs * rhs);
+    expected_lhs(MUL) = d_lhs * d_rhs;
+
+    actual_lhs(DIV) = Kokkos::Experimental::cast_from_half<double>(lhs / rhs);
+    expected_lhs(DIV) = d_lhs / d_rhs;
+
+    actual_lhs(NEG) = Kokkos::Experimental::cast_from_half<double>(!lhs);
+    expected_lhs(NEG) = !d_lhs;
+
+    actual_lhs(AND) = Kokkos::Experimental::cast_from_half<double>(half_t(0) && lhs);
+    expected_lhs(AND) = double(0) && d_lhs;
+
+    actual_lhs(OR) = Kokkos::Experimental::cast_from_half<double>(lhs || half_t(1));
+    expected_lhs(OR) = d_lhs || double(1);
+  }
+
+  inline
+  void run_on_device() {
+    Kokkos::parallel_for(
+            "Test::Functor_TestHalfOperators",
+            Kokkos::RangePolicy<>(0,1),
+            *this);
```

Comment:
I would stick this directly in the constructor.

Reply:
Also we might want to be explicit about the execution space.

Reply:
I was only sticking this into the default test executable which uses default execution space.